### PR TITLE
AccessControl for NativeAudio

### PR DESCRIPTION
**Pull request by noslouch**
iOS Safari and the Audio element do not play nicely when it comes to starting audio outside of a direct click event.

the `AccessControl` util serves as something of a mutex so that we can generate and pass around a "magic" audio element that has been opened by the power the human touch and into which we can swap audio sources for programmatic playback.

This might be missing some tests. Would love some feedback on any use cases that need to be verified.

### CODE REVIEW

**Comment by jkeen** [comment]
This looks good for the bug it fixes, but I realized it breaks one idea that we designed ember-hifi's architecture to work with: while only one sound can be playing at a time, multiple sounds should be able to be loaded on the page and each should still be able to set/get position, know its duration, etc. Basically each should be able to have mini controls that interact with the sound itself, while the service enforces that only one can be playing at a time.

Even without having separate playback controls for each sound, I wonder what would happen with this logic if you: 
 - play sound 1 
 - at `15s` into sound 1, start playing sound 2
 - resume sound 1

Will sound 1 remember to start at `15s`? I'd guess it would start at the beginning again.

I think we should be able to fake this out with a little trickery by doing something like `audio.cloneNode()` when we're transferring access, so inactive sounds have a cloned copy of the shared resource, and then when we allow access to the shared audio element we make sure to set some attributes.

**Comment by jkeen** [comment]
This needs a test verifying that if hooking into the `audio-ended` event on the service and then triggering a play on another sound in that event that the sound still uses the shared element.

In working on this a bit tonight, I spotted an error where it doesn't look like the attributes are actually being restored after switching between a shared element and the dummy/internal element, and in fixing that am now dealing with other errors. The slog continues.

Also, now Native Audio only works with a shared audio element (even though all other browsers besides iOS don't need this strategy), and the hifi service passes it in every time. We might want to modify that.

File: addon/hifi-connections/native-audio.js [hunk 1]
```diff
@@ -88,6 +95,54 @@ let Sound = BaseSound.extend({
...
+
+  requestControl() {
+    if (!this.get('sharedAudioAccess')) {
+      return;
+    }
+
+    this.get('sharedAudioAccess').requestControl(this);
+
+    let sharedElement     = this.audioElement();
+    let internalElement   = this.get('audioElement');
```
Comment (noslouch) [inline addon/hifi-connections/native-audio.js]: This should probably be _audioElement

**Comment by noslouch** [comment]
@jkeen, I'm sad to say but we've got a fundamental flaw in our premise here: https://jsbin.com/gimokokozu/edit?js,console

`cloneNode` doesn't copy attributes like `currentTime` and `volume` :/

**Comment by jkeen** [comment]
@noslouch 😭

**Comment by jkeen** [comment]
Welp, I guess our existing tests suck.

**Comment by jkeen** [comment]
4

**Comment by noslouch** [comment]
part of the problem is that we're looking at the computed `'position'`, which is updated by the service polling the `currentSound`. it's a bit circular, but we should probably check the value of `_currentPosition`, because that is what will become the source of truth for that attr

**Comment by noslouch** [comment]
ok, got a failing test on `'position'`

**Comment by noslouch** [comment]
but I'm still getting caught up on setting a property on a destroyed element in the hifi service test 😢

**Comment by jkeen** [comment]
@noslouch This is the same sort of thing I ran into when `SoundCache` was just an `Ember.Object`, in that every second test run it would somehow stick around between tests and wreak havoc. So I changed it to a service and stubbed it out.

I thought this was an issue with the shared audio control and was trying to reset it in the after hook of the tests, but that didn't seem to work either. Intermittent test failures are the worst. I wonder if we have to stub that thing too?

**Comment by noslouch** [comment]
looks like soundcache is already stubbed in the service unit test

**Comment by noslouch** [comment]
@jkeen could use a review. I think we're close.

**Comment by jkeen** [comment]
@noslouch Once this passes I think we should be good to merge and tag if your tests using a real iPhone on wnyc.org work as expected

**Pull request merged**